File: ruby/src/training_data_formatter.rb
# frozen_string_literal: true
# rbs_inline: enabled

require_relative 'lib/format'

# Reads a CSV training-data file, converts it through Lib::Format, and writes
# the result as a timestamped JSON file under the requested directory.
class TrainingDataFormatter
  include ::Lib::Format

  # @rbs path_to_csv_training_data: String
  # @rbs path_to_json_training_data: String
  # @rbs return: void
  def self.run(path_to_csv_training_data:, path_to_json_training_data:)
    new(path_to_csv_training_data:, path_to_json_training_data:).run
  end

  # @rbs path_to_csv_training_data: String
  # @rbs path_to_json_training_data: String
  # @rbs return: void
  def initialize(path_to_csv_training_data:, path_to_json_training_data:)
    @path_to_csv_training_data  = path_to_csv_training_data
    @path_to_json_training_data = path_to_json_training_data
  end

  # @rbs return: void
  def run
    File.write(filename, to_json(path_to_csv_training_data:))
  end

  private

  attr_reader :path_to_csv_training_data, :path_to_json_training_data

  # @rbs path_to_json_training_data: String
  # @rbs return: String
  def filename
    File.join(path_to_json_training_data, "training_data_#{Time.now.strftime('%F%T').gsub(/[:-]/, '')}.json")
  end
end

File: python/src/training_data_formatter.py
import datetime
import os
import sys

sys.path.append('./src/lib')

from format import Format  # noqa: E402


class TrainingDataFormatter(Format):
    """Reads a CSV training-data file, converts it through Format, and writes
    the result as a timestamped JSON file under the requested directory."""

    @classmethod
    def run(cls, path_to_csv_training_data: str,
            path_to_json_training_data: str) -> None:
        cls(path_to_csv_training_data, path_to_json_training_data)._run()

    def __init__(self, path_to_csv_training_data: str,
                 path_to_json_training_data: str) -> None:
        self._path_to_csv_training_data = path_to_csv_training_data
        self._path_to_json_training_data = path_to_json_training_data

    # private

    def _run(self) -> None:
        with open(self._filename(), 'w') as f:
            f.write(self.to_json(self._path_to_csv_training_data))

    def _filename(self) -> str:
        return os.path.join(
            self._path_to_json_training_data,
            f'training_data_{datetime.datetime.now():%Y%m%d%H%M%S}.json',
        )
